reading_comprehension/qanet_encoder.py

```diff
@@ -0,0 +1,345 @@
+    input_dim : ``int``, required.
+        The input dimension of the encoder.
+    hidden_dim : ``int``, required.
+        The hidden dimension used for convolution output channels, multi-head attention output
+        and the final output of feedforward layer.
+    attention_projection_dim : ``int``, required.
+        The dimension of the linear projections for the self-attention layers.
+    feedforward_hidden_dim : ``int``, required.
+        The middle dimension of the FeedForward network. The input and output
+        dimensions are fixed to ensure sizes match up for the self attention layers.
+    num_convs_per_block: ``int``, required.
+        The number of convolutions in each block.
+    conv_kernel_size: ``int``, required.
+        The kernel size for convolution.
+    num_attention_heads : ``int``, required.
+        The number of attention heads to use per layer.
+    use_positional_encoding: ``bool``, optional, (default = True)
+        Whether to add sinusoidal frequencies to the input tensor. This is strongly recommended,
+        as without this feature, the self attention layers have no idea of absolute or relative
+        position (as they are just computing pairwise similarity between vectors of elements),
+        which can be important features for many tasks.
+    dropout_prob : ``float``, optional, (default = 0.1)
+        The dropout probability for the feedforward network.
+    layer_dropout_undecayed_prob : ``float``, optional, (default = 0.1)
+        The initial dropout probability for layer dropout, and this might decay w.r.t the depth
+        of the layer. For each mini-batch, the convolution/attention/ffn sublayer is
+        stochastically dropped according to its layer dropout probability.
+    attention_dropout_prob : ``float``, optional, (default = 0)
+        The dropout probability for the attention distributions in the attention layer.
+    """
+
+    def __init__(self,
+                 input_dim: int,
+                 hidden_dim: int,
+                 attention_projection_dim: int,
+                 feedforward_hidden_dim: int,
+                 num_convs: int,
+                 conv_kernel_size: int,
+                 num_attention_heads: int,
+                 use_positional_encoding: bool = True,
+                 dropout_prob: float = 0.1,
+                 layer_dropout_undecayed_prob: float = 0.1,
+                 attention_dropout_prob: float = 0) -> None:
+        super(QANetEncoderBlock, self).__init__()
+
+        assert input_dim == hidden_dim
+
+        self._use_positional_encoding = use_positional_encoding
+
+        self._conv_norm_layers: List[LayerNorm] = []
+        self._conv_layers: List[DepthwiseSeparableConv] = []
+
+        for conv_idx in range(num_convs):
+            conv_norm = LayerNorm(hidden_dim)
+            # call add_module because this module is not a member variable of this class
+            self.add_module(f"conv_layer_norm_{conv_idx}", conv_norm)
```

Comment:
I think there's a `ModuleList` that you can use to make this a little cleaner; not a big deal, though.